File: day4/python/part_1.py
from typing import List
from utils import Grid, read_lines


def count_xmas(grid: List[List[str]]) -> int:
    word = "XMAS"
    directions = [(-1, 0), (1, 0), (0, -1), (0, 1), (-1, -1), (-1, 1), (1, -1), (1, 1)]
    grid_obj = Grid(grid)  
    count = 0

    def valid(r: int, c: int, dr: int, dc: int) -> bool:
        for i in range(len(word)):
            nr, nc = r + dr * i, c + dc * i
            if grid_obj.get(nr, nc) != word[i]:
                return False
        return True

    for r in range(grid_obj.rows):
        for c in range(grid_obj.cols):
            for dr, dc in directions:
                if valid(r, c, dr, dc):
                    count += 1
    return count



grid = [list(line.strip()) for line in read_lines("input.txt")]
print(count_xmas(grid))

File: day4/typescript/part_1.ts
import * as fs from 'fs';

type Direction = [number, number];

function countXmas(grid: string[][]): number {
    const word = "XMAS";
    const directions: Direction[] = [
        [-1, 0], [1, 0], [0, -1], [0, 1], 
        [-1, -1], [-1, 1], [1, -1], [1, 1]
    ];
    const rows = grid.length;
    const cols = grid[0].length;
    let count = 0;

    function valid(r: number, c: number, dr: number, dc: number): boolean {
        for (let i = 0; i < word.length; i++) {
            const nr = r + dr * i;
            const nc = c + dc * i;
            if (nr < 0 || nr >= rows || nc < 0 || nc >= cols || grid[nr][nc] !== word[i]) {
                return false;
            }
        }
        return true;
    }

    for (let r = 0; r < rows; r++) {
        for (let c = 0; c < cols; c++) {
            for (const [dr, dc] of directions) {
                if (valid(r, c, dr, dc)) {
                    count++;
                }
            }
        }
    }

    return count;
}

const grid: string[][] = fs.readFileSync('input.txt', 'utf-8')
    .split('\n')
    .map(line => line.split(''));

console.log(countXmas(grid)); 
